File: test/lightningcss/bundle_test.rb
# frozen_string_literal: true

require "test_helper"

module LightningCSS
  class BundleTest < Minitest::Spec
    ENTRY = File.expand_path("../fixtures/bundle/entry.css", __dir__)
    MISSING = File.expand_path("../fixtures/bundle/missing.css", __dir__)
    WARNED = File.expand_path("../fixtures/bundle/warned.css", __dir__)
    BROKEN = File.expand_path("../fixtures/bundle/broken.css", __dir__)
    UNREADABLE = File.expand_path("../fixtures/bundle/nested/broken.css", __dir__)
    IMPORTED = File.expand_path("../fixtures/bundle/nested/warned.css", __dir__)
    WARNING = "'deep' is not recognized as a valid pseudo-class. Did you mean '::deep' (pseudo-element) or is this a typo? at #{IMPORTED}:0:9".freeze

    test "resolves the imports a stylesheet was written with, in the order it imported them" do
      code = LightningCSS.bundle(ENTRY, minify: true).code

      assert_equal ":root{--brand:red}.layout{display:grid}.entry{color:var(--brand)}", code
    end

    test "narrows everything it bundled by a scope" do
      code = LightningCSS.bundle(ENTRY, minify: true, scope: "[s]").code

      assert_equal ":root[s]{--brand:red}.layout[s]{display:grid}.entry[s]{color:var(--brand)}", code
    end

    test "raises a bundle error when a file it was sent to read is not there" do
      error = assert_raises(LightningCSS::BundleError) { LightningCSS.bundle(MISSING) }

      assert_equal "No such file or directory (os error 2)", error.message
    end

    test "a bundle error is an error" do
      assert_operator LightningCSS::BundleError, :<, LightningCSS::Error
    end

    test "a transformer carries its options into a bundle" do
      code = Transformer.new(minify: true).bundle(ENTRY).code

      assert_equal ":root{--brand:red}.layout{display:grid}.entry{color:var(--brand)}", code
    end

    test "renames every name it bundled, and reports the ones the entry wrote" do
      result = LightningCSS.bundle(ENTRY, css_modules: { pattern: "bundled-[local]" }, minify: true)

      assert_equal ":root{--brand:red}.bundled-layout{display:grid}.bundled-entry{color:var(--brand)}", result.code
      assert_equal({ "entry" => "bundled-entry" }, result.exports)
    end

    test "hashes every file it bundled on its own, so two of them never collide" do
      code = LightningCSS.bundle(ENTRY, css_modules: true, minify: true).code

      entry = code[/\.(\w+)_entry\{/, 1]
      layout = code[/\.(\w+)_layout\{/, 1]

      refute_nil entry
      refute_nil layout
      refute_equal entry, layout
    end

    test "reports no exports when it was not asked to compile a CSS module" do
      assert_nil LightningCSS.bundle(ENTRY, minify: true).exports
    end

    test "reports what it kept but did not understand, and which file wrote it" do
      result = LightningCSS.bundle(WARNED, minify: true)

      assert_equal [WARNING], result.warnings
      assert_predicate result, :warnings?
    end

    test "keeps the rule it warned about" do
      assert_includes LightningCSS.bundle(WARNED, minify: true).code, ".a:deep(.b){color:red}"
    end

    test "reports none for a bundle it fully understood" do
      result = LightningCSS.bundle(ENTRY, minify: true)

      assert_empty result.warnings
      refute_predicate result, :warnings?
    end

    test "raises a parse error for CSS it could not read, wherever it imported it from" do
      error = assert_raises(LightningCSS::ParseError) { LightningCSS.bundle(BROKEN) }

      assert_equal "Invalid empty selector at #{UNREADABLE}:0:1", error.message
    end

    test "refuses a filename, naming every file it reads by the path it read it from" do
      error = assert_raises(LightningCSS::OptionError) { LightningCSS.bundle(ENTRY, filename: "x.css") }

      assert_equal "filename is not an option for a bundle", error.message
    end

    test "refuses a filename a transformer was built with, the same way" do
      error = assert_raises(LightningCSS::OptionError) { Transformer.new(filename: "x.css").bundle(ENTRY) }

      assert_equal "filename is not an option for a bundle", error.message
    end
  end
end

File: rust/tests/bundle_test.rs
use lightningcss_ffi::{lightningcss_bundle, lightningcss_result_free, LightningCssErrorCode};

use std::ffi::{CStr, CString};

const ENTRY: &str = concat!(env!("CARGO_MANIFEST_DIR"), "/../test/fixtures/bundle/entry.css");
const MISSING: &str = concat!(env!("CARGO_MANIFEST_DIR"), "/../test/fixtures/bundle/missing.css");
const BROKEN: &str = concat!(env!("CARGO_MANIFEST_DIR"), "/../test/fixtures/bundle/broken.css");

fn bundle(path: &str, options: &str) -> Result<String, String> {
  let path = CString::new(path).unwrap();
  let options = CString::new(options).unwrap();

  let result = unsafe { lightningcss_bundle(path.as_ptr(), options.as_ptr()) };

  let outcome = if result.error.is_null() {
    Ok(unsafe { CStr::from_ptr(result.value) }.to_str().unwrap().to_string())
  } else {
    Err(unsafe { CStr::from_ptr(result.error) }.to_str().unwrap().to_string())
  };

  unsafe { lightningcss_result_free(result) };

  outcome
}

#[test]
fn bundle_resolves_the_imports_a_stylesheet_was_written_with() {
  let answer = bundle(ENTRY, r#"{"minify":true}"#);

  assert_eq!(
    answer,
    Ok(r#"{"code":":root{--brand:red}.layout{display:grid}.entry{color:var(--brand)}","warnings":[]}"#.to_string())
  );
}

#[test]
fn bundle_narrows_everything_it_read_by_a_scope() {
  let answer = bundle(ENTRY, r#"{"minify":true,"scope":"[s]"}"#);

  assert_eq!(
    answer,
    Ok(
      r#"{"code":":root[s]{--brand:red}.layout[s]{display:grid}.entry[s]{color:var(--brand)}","warnings":[]}"#
        .to_string()
    )
  );
}

#[test]
fn bundle_answers_an_error_for_a_file_that_is_not_there() {
  assert_eq!(
    bundle(MISSING, "{}"),
    Err("No such file or directory (os error 2)".to_string())
  );
}

fn bundle_code(path: &str, options: &str) -> LightningCssErrorCode {
  let path = CString::new(path).unwrap();
  let options = CString::new(options).unwrap();

  let result = unsafe { lightningcss_bundle(path.as_ptr(), options.as_ptr()) };
  let answer = result.code;

  unsafe { lightningcss_result_free(result) };

  answer
}

#[test]
fn a_file_it_cannot_read_carries_the_bundle_code() {
  assert_eq!(bundle_code(MISSING, "{}"), LightningCssErrorCode::Bundle);
}

#[test]
fn css_it_cannot_parse_carries_the_parse_code_wherever_it_was_imported_from() {
  assert_eq!(bundle_code(BROKEN, "{}"), LightningCssErrorCode::Parse);
}
